tree.rb

```diff
@@ -0,0 +1,123 @@
+# ! MINOR BUGS: 
+# 1) :text is being deleted from the <ul> node. I know this has to do with 
+# the special gsub for <li> children but not specifically why it's happening.
+# 2) Spacing error on some of the other :text fields. Specifically the 'outer-div'
+
+# Create a DOM Tree Class
+
+# First thing is to create the Node struct
+Node = Struct.new(:name, :text, :classes, :id, :children, :parent)
+
+class DOMTree
+	attr_reader :file
+	attr_accessor :head
+
+	def initialize(file_name)
+		# gsub(/<!.*?>/, "") Removes <! doctype > from the document
+		# gsub(/>\s*</,"><") Removes all the whitespace from between > & < characters
+		# gsub(/\n/, "") Removes the newline characters from the document
+		@file = load(file_name).read.gsub(/\n/, "").gsub(/>\s*</,"><").gsub(/<!.*?>/, "")
+		@head = Node.new("Document", @file)
+	end
+
+	def load(file_name)
+		File.open(file_name, "r")
+	end
+
+	# Recursively get the children for each node. Returns nil
+	# if there are no more children.
+	def create_tree
+		queue = [@head]
+		while !queue.empty?
+			current_node = queue.pop
+			if has_children?(current_node.text)
+				current_node.children = []
+				create_children(current_node, queue)
+			else
+				current_node.text = current_node.text.strip
+			end
+		end
+	end
+
+	def create_children(node, queue)
+		while has_children?(node.text)
+			first_tag = get_first_tag(node.text)
+			child_name = get_name(first_tag)
+			child_class = (get_class(first_tag) == false ? nil :  get_class(first_tag))
+			child_id = (get_id(first_tag) == false ? nil :  get_id(first_tag))
```

Comment:
Close, but I'd do more of a "tell don't ask" with this -- just say `child_id = get_id(first_tag)` and let `get_id` handle the conditionality. Same for the rest.